### PR TITLE
Fix TextGeneration and Text2TextGeneration pipeline issue with return_dict_in_generate

### PR DESCRIPTION
**Pull request by tokestermw**
# What does this PR do?

When running `return_dict_in_generate=True` in pipelines, the `generate` method errors out, e.g.

```python
        generate_kwargs["min_length"] = generate_kwargs.get("min_length", self.model.config.min_length)
        generate_kwargs["max_length"] = generate_kwargs.get("max_length", self.model.config.max_length)
        self.check_inputs(input_length, generate_kwargs["min_length"], generate_kwargs["max_length"])
        output_ids = self.model.generate(**model_inputs, **generate_kwargs)
E       AttributeError: 'GreedySearchEncoderDecoderOutput' object has no attribute 'shape'
```

Reproducible code:

```python
from transformers import pipeline
generator = pipeline('text-generation', 'gpt2')
generator('hello', return_dict_in_generate=True)
```

Fix is to add a check if `return_dict_in_generate=True`, then let `generated_sequence = generated_sequence.sequences`.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

@gante, @Narsil

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: tests/pipelines/test_pipelines_text2text_generation.py [hunk 1]
```diff
@@ -124,3 +124,31 @@ def test_small_model_tf(self):
         # do_sample=False necessary for reproducibility
         outputs = generator("Something there", do_sample=False)
         self.assertEqual(outputs, [{"generated_text": ""}])
+
+    @require_torch
+    def test_return_dict_in_generate(self):
+        greedy_search_inference_config = {"do_sample": False, "num_return_sequences": 1, "num_beams": 1}
+        pipe = pipeline(task="text2text-generation", model="patrickvonplaten/t5-tiny-random")
+        pipe("hello", return_dict_in_generate=True, **greedy_search_inference_config)
```
Comment (Narsil) [inline tests/pipelines/test_pipelines_text2text_generation.py:132]: Do you mind adding a check for those returned values?

File: src/transformers/pipelines/text2text_generation.py [hunk 2]
```diff
@@ -185,6 +185,10 @@ def _forward(self, model_inputs, **generate_kwargs):
         generate_kwargs["max_length"] = generate_kwargs.get("max_length", self.model.config.max_length)
         self.check_inputs(input_length, generate_kwargs["min_length"], generate_kwargs["max_length"])
         output_ids = self.model.generate(**model_inputs, **generate_kwargs)
+        if generate_kwargs.get("return_dict_in_generate"):
+            # Instance check of type unions do not work until Python 3.10, e.g. `isinstance(generated_sequence, GenerateOutput):`
+            # https://peps.python.org/pep-0604/#isinstance-and-issubclass
```
Comment (Narsil) [inline src/transformers/pipelines/text2text_generation.py:190]: This comment is unnecessary imo. The condition you're looking at is the correct one.

**Review by Narsil — COMMENTED** [review]
Hi thank you for this PR. Do you mind sharing in which context you need to use this return changing flag?

File: tests/pipelines/test_pipelines_text2text_generation.py (hunk 1 above)
Comment (tokestermw) [reply tests/pipelines/test_pipelines_text2text_generation.py]: done!

File: src/transformers/pipelines/text2text_generation.py (hunk 2 above)
Comment (tokestermw) [reply src/transformers/pipelines/text2text_generation.py]: done!

**Comment by tokestermw** [comment]
> Hi thank you for this PR. Do you mind sharing in which context you need to use this return changing flag?

Currently I use the scores for a custom stopping criterion (e.g. stop when cumulative probs are > x).

**Comment by gante** [comment]
@tokestermw interesting! `pipeline` only return the output sequence, but to store the `scores` internally in `.generate()`, `return_dict_in_generate` is indeed needed. Is this what is happening in your use case?

**Comment by tokestermw** [comment]
@gante yes that's right! to access the `scores` here: https://github.com/huggingface/transformers/blob/42b60f8b02941b0c40c42e150a101eb372c3856e/src/transformers/generation/stopping_criteria.py#L37

**Comment by Narsil** [comment]
> Currently I use the scores for a custom stopping criterion (e.g. stop when cumulative probs are > x).

Shouldn't that be done with a custom `StoppingCriteria` ? 

This PR a small enough to be ok anything, just wondering if there's not a "cleaner" way for  you to solve your issue.

**Comment by tokestermw** [comment]
@Narsil we do use a custom `StoppingCriteria`, and use it inside pipelines.

But yeah arguably custom stuff should maybe be done outside of pipelines.

Sidenote is that we can't currently pass in tokenizer args like `truncation=True` in pipelines, so we've had to either write a custom pipeline, or rewrite to not use pipelines :)

**Comment by Narsil** [comment]
> Sidenote is that we can't currently pass in tokenizer args like truncation=True in pipelines, so we've had to either write a custom pipeline, or rewrite to not use pipelines :)

We could definitely add `tokenizer_kwargs` to enable all the args you want on the tokenizer.

Another way of doing it would be by subclassing

https://huggingface.co/docs/transformers/v4.26.0/en/add_new_pipeline#how-to-create-a-custom-pipeline

And then using `pipeline(...., pipeline_class=MyPipelineClass)` to use it instead of the default one.
Then you can add all the fancy logic you need.

The docs also refer on how to share it !

**Comment by tokestermw** [comment]
Thanks @Narsil ! i can take a look at `tokenizer_kwargs` in a separate PR, please let me know if anything else is needed in this PR

**Comment by Narsil** [comment]
Seems good, just don't pass all `tokenizer_kwargs` as a generic `**kwargs`. 

Like let's capture `pipeline(...tokenizer_kwargs={"truncation": True})` so that it cannot clash with `generate_kwargs` (inconveninent but necessary, `max_length` is an argument for both)

**Comment by ArthurZucker** [comment]
cc @echarlaix, potentialy including `tokenizer_kwargs` is planned 😉

**Pull request closed without merging**